src/s3/explorer/s3BucketNode.ts

```diff
@@ -0,0 +1,113 @@
+import { AWSTreeNodeBase } from '../../shared/treeview/nodes/awsTreeNodeBase'
+import { ErrorNode } from '../../shared/treeview/nodes/errorNode'
+import { LoadMoreNode } from '../../shared/treeview/nodes/loadMoreNode'
+import { PlaceholderNode } from '../../shared/treeview/nodes/placeholderNode'
+import { makeChildrenNodes } from '../../shared/treeview/treeNodeUtilities'
+import { S3FolderNode } from './s3FolderNode'
+import { S3MoreResultsNode } from './s3MoreResultsNode'
+import { S3NodeCache } from './s3NodeCache'
+import { S3FileNode } from './s3FileNode'
+import { localize } from '../../shared/utilities/vsCodeUtils'
+
+/**
+ * Represents an S3 bucket that may contain folders and/or objects.
+ */
+export class S3BucketNode extends AWSTreeNodeBase implements AWSResourceNode, LoadMoreNode {
+    private cache: S3NodeCache
+
+    public constructor(public readonly bucket: Bucket, private readonly s3: S3Client) {
+        super(bucket.name, vscode.TreeItemCollapsibleState.Collapsed)
+        this.tooltip = bucket.name
+        this.iconPath = {
+            dark: vscode.Uri.file(ext.iconPaths.dark.s3.bucket),
+            light: vscode.Uri.file(ext.iconPaths.light.s3.bucket),
+        }
+        this.contextValue = 'awsS3BucketNode'
+        this.cache = new S3NodeCache(new S3MoreResultsNode(this))
+    }
+
+    public async getChildren(): Promise<AWSTreeNodeBase[]> {
+        return await makeChildrenNodes({
+            getChildNodes: async () => {
+                // Consider locking to avoid possible race conditions
+                if (!this.initialChildrenLoaded()) {
+                    return this.loadInitialChildren()
+                }
+
+                return this.getExistingChildren()
+            },
+            getErrorNode: async (error: Error) =>
+                new ErrorNode(this, error, localize('AWS.explorerNode.s3.error', 'Error loading S3 resources')),
+            getNoChildrenPlaceholderNode: async () =>
+                new PlaceholderNode(this, localize('AWS.explorerNode.s3.noFiles', '[No Files found]')),
+        })
+    }
+
+    private initialChildrenLoaded(): boolean {
+        return !this.cache.isPristine
+    }
+
+    private async loadInitialChildren(): Promise<AWSTreeNodeBase[]> {
+        return this.loadMoreChildren()
+    }
+
+    private async getExistingChildren(): Promise<AWSTreeNodeBase[]> {
+        return this.cache.nodes
+    }
+
+    public async loadMoreChildren(): Promise<AWSTreeNodeBase[]> {
+        const response = await this.s3.listObjects({
+            bucketName: this.bucket.name,
+            continuationToken: this.cache.continuationToken,
+        })
+
+        const newFolders = response.folders.map(folder => new S3FolderNode(this.bucket, folder, this.s3))
+        const newFiles = response.files.map(file => new S3FileNode(this.bucket, file, this.s3))
+        this.cache.appendItems(newFolders, newFiles, response.continuationToken)
```

Comment:
Nit: Move this logic to the cache (just send the response object without filtering) since this is duplicated in the folder node. (This didn't get added in the last review for some reason...)

Reply:
I think there's a more dramatic de-duplication that should be done between Folder and Bucket considering how similar they are to one another. I had tried to add stuff to an abstract class, but it was creating a circular dependency so I had scrapped it. This is one of the areas I wasn't satisfied with. Will look again.

Reply:
I kept `Bucket` and `File` as separate *(similar)* nodes. However, I extracted out the caching and "load more..." functionality which was what I was worried about.